lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+//===- HWSpecialize.cpp - hw module specialization ------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transform performs specialization of parametric hw.module's.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Support/ValueMapper.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+
+namespace {
+
+using InstanceParameters = llvm::DenseMap<hw::HWModuleOp, ArrayAttr>;
+
+// Generates a module name by composing the name of 'moduleOp' and the set of
+// provided 'parameters'.
+static std::string generateModuleName(hw::HWModuleOp moduleOp,
+                                      ArrayAttr parameters) {
+  assert(parameters.size() != 0);
+  std::string name = moduleOp.getName().str();
+  for (auto param : parameters) {
+    auto paramAttr = param.cast<ParamDeclAttr>();
+    int64_t paramValue = paramAttr.getValue().cast<IntegerAttr>().getInt();
+    name += "_" + paramAttr.getName().str() + "_" + std::to_string(paramValue);
+  }
+
+  // Uniqueness check
+  auto parentModule = moduleOp->getParentOfType<ModuleOp>();
+  int uniqueCntr = 0;
+  std::string uniqueName = name;
+  while (parentModule.lookupSymbol(uniqueName))
```

Comment:
Isn't `lookupSymbol()` a linear lookup? If so, please avoid by building a cache at the beginning of the pass and maintaining it.

Reply:
I believe it is, and Morten if you hadn't seen it, I think this is the cache you can use: https://github.com/llvm/circt/blob/main/include/circt/Dialect/HW/HWSymCache.h

Reply:
IIRC that has to be frozen before doing lookups, but since this pass is adding modules it has to be written to. This is a relatively common pattern so we should probably get one. Actually, name uniquing could probably be part of it.

Reply:
Modified to use SymbolCache. Factored out the module-based initialization found in MSFTPasses into the SymbolCache itself, as well as an option to unfreeze the cache, to allow for mutation.

Reply:
@darthscsi Is adding an `unfreeze` capability to `hw::SymbolCache` kosher?

Reply:
Great refactoring! This should probably be done in a separate PR. It might also be a good idea to put the `unfreeze` functionality in a subclass.

Reply:
I can pull out the unfreezing into a `MutableSymCache` class. My take on `SymCache` unfreezing: since there are no synchronization capabilities in `SymCache` for making it thread safe, I don't see an issue with unfreezing it, since `SymCache` safety is already 100% down to the programmer (hence the asserts in set/getDefinition)).

Reply:
I don't understand... SymCache is supposed to be constructed in a single thread, frozen, then used by many threads. So it doesn't need any synchronization capabilities for that purpose.

Reply:
Sorry, let me elaborate; it would need synchronization capabilities if it's expected to function in a multithreaded context, where unfreezings may occur.

Reply:
Ah. Yes. By making it a subclass, however, users can just use the parent class and be assured of the original behavior. If someone needs mutability, you're correct -- synchronization is entirely on them.